src/env.rs

```diff
@@ -17,35 +24,46 @@ pub fn get_pam_env(handle: &mut PamHandle) -> Option<PamEnvList> {
 }
 
 impl PamEnvList {
-    pub fn to_vec(&mut self) -> Vec<(String, String)> {
-        let mut vec = Vec::new();
-
-        let mut idx = 0;
-        loop {
-            let env_ptr: *const *const c_char = unsafe { self.ptr.offset(idx) };
-            if unsafe { !(*env_ptr).is_null() } {
-                idx += 1;
-
-                let env = unsafe { CStr::from_ptr(*env_ptr) }.to_string_lossy();
-                let split: Vec<_> = env.splitn(2, '=').collect();
-
-                if split.len() == 2 {
-                    // Only add valid env vars (contain at least one '=')
-                    vec.push((split[0].into(), split[1].into()));
-                }
+    pub fn iter(&self) -> IterPamEnv {
+        IterPamEnv {
+            envs: self,
+            idx: 0,
+            ended: false,
+        }
+    }
+
+    pub fn as_ptr(&self) -> *const *const c_char {
+        self.ptr
+    }
+}
+
+impl<'a> Iterator for IterPamEnv<'a> {
+    type Item = &'a CStr;
+
+    fn next(&mut self) -> Option<&'a CStr> {
+        if self.ended {
+            return None;
+        }
+
+        unsafe {
```

Comment:
I would prefer, I we restrict the `unsafe` as much as possible, e.g., to the call to `offset` and the dereferencing when creating the `CStr`.